cranelift/interpreter/src/state.rs

```diff
@@ -0,0 +1,117 @@
+//! Cranelift instructions modify the state of the machine; the [State] trait describes these
+//! ways this can happen.
+use cranelift_codegen::ir::condcodes::{FloatCC, IntCC};
+use cranelift_codegen::ir::{FuncRef, Function, Type, Value};
+use cranelift_entity::PrimaryMap;
+use smallvec::SmallVec;
+use thiserror::Error;
+
+/// This trait manages the state necessary to interpret a single Cranelift instruction.
```

Comment:
Yup... this was left a bit unfinished because I was (and am still) debating whether this should be a tree of traits instead of a single level. For example, `State` could contain `fn iflags() -> impl Flags<IntCC>` and then separately I could define `has/set/clear` for `Flags`. Why I haven't done this? I'm concerned that, since `impl Flags` is likely not correct (though it expresses what I mean), I will be caught up in `dyn` issues, possible lifetime issues, mutable vs. non-mutable issues, etc. Do you have an opinion on this?

Reply:
FYI, traits can't contain `-> impl Trait` methods, you would have to have an associated type that is returned instead:

```
trait State {
    type IFlags: IFlagsState;

    fn iflags(&mut self) _> &mut Self::Iflags;

    // etc...
}
```

I generally don't think it is worth it to do this sort of thing unless there is code that will be generic over just `IFlagsState` and not the whole `State`, which I suspect isn't the case here.

Reply:
👍 I documented it better and left it as-is.